### PR TITLE
Bug 1815638 : disable permission prompt pre checking

### LINKED ISSUE #26577: [Bug]: Keyboard is show over the QR Scanner on the first launch
### Steps to reproduce

1. Install Firefox Nightly.
2. Use a fresh profile.
3. Go to the search bar and tap on QR Scanner.
4. Click "While using the app" when camera access prompt is shown.
5. Observe.

### Expected behaviour

The Keyboard should not be present when scanning a QR Code.

### Actual behaviour

The keyboard is shown over the QR Scanner screen.

### Device name

Huawei P40 Lite

### Android version

Android 10

### Firefox release type

Firefox Nightly

### Firefox version

105.0a1

### Device logs

_No response_

### Additional information

The keyboard is also working and the user can write in the Search Bar while in QR Scanner Screen.

https://user-images.githubusercontent.com/103257761/185905436-149a863c-93e5-4dd9-aa9d-cfdfc6600e97.mp4

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21353)

### LINKED ISSUE #27177: Camera permission flow do not handle : `Ask Every Time` permission
### Steps to reproduce

1. On android 11 or later
2. Go the address bar
3. Use the scanner feature
4. When the permission prompt open select : Only this time
5. Close the application
6. Force it to stop (Long press on icon / Info / Force stop)
7. Reopen and select the scanner feature again

### Expected behaviour

The permission prompt should open and allow to grant the permission.

### Actual behaviour

Another dialog window open which prompt the user to grant the permission. It offers two actions : ignore or bring the user to the application settings.
But when going to the application settings the camera permission appear correctly as  `Ask every time`.

### Device name

Google Pixel 4a (5G) on CalyxOS

### Android version

Android 12

### Firefox release type

Firefox

### Firefox version

105.1.0 

### Device logs

_No response_

### Additional information

First some of the label I'm quoting might be wrong since I'm running in French on my phone sorry ^^.

As mentioned, I believe the `Ask every time` option appeared with Android 11 ([Source](https://support.google.com/android/answer/9431959?hl=en))

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21674)

### PR DESCRIPTION
**Pull request by Timshel**
Change the camera logic to always prompt the user and only rely on a setting to memorize if we are in the `Denied Forever` state.

Note: logic could be implemented without storing a state but it would mean opening the permission dialog change when we reached the `Denied Forever` state just after the user refused to grant the permission (a bit obnoxious I believe).

Switched to using `ActivityResultLauncher` to request the permission as mentioned in (https://github.com/mozilla-mobile/fenix/issues/19920) since it made code factorization easier.

It could be made cleaner by modifying the `QrFeature` component to change the signature of the parameter `onNeedToRequestPermissions` and the function `onPermissionsResult`.

As a side effect of working on the same code it include logic from https://github.com/mozilla-mobile/fenix/pull/27062 and fix #26577 

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #27177

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt [hunk 1]
```diff
@@ -343,19 +352,7 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
...
 
-            if (requireContext().settings().shouldShowCameraPermissionPrompt) {
-                qrFeature.get()?.scan(binding.searchWrapper.id)
-            } else {
-                if (requireContext().isPermissionGranted(Manifest.permission.CAMERA)) {
-                    qrFeature.get()?.scan(binding.searchWrapper.id)
-                } else {
-                    interactor.onCameraPermissionsNeeded()
-                    resetFocus()
-                    view.hideKeyboard()
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:354]: Drive-by review: Are we certain, we don't need to dismiss the keyboard as well?

Comment (Timshel) [reply app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:354]: Looking at it there is a specific call to close the keyboard when dismissing the dialog : 
https://github.com/mozilla-mobile/fenix/blob/5cce4b5f15ffcbfa70079318f284c67de6d8bb6b/app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt#L548

Comments of `hideDeviceKeyboard` method reference what appear to be a different use case for the dialog, will look at it.

**Comment by Timshel** [comment]
Made some change for the permission dialog not to rely the shared dismiss logic in  https://github.com/mozilla-mobile/fenix/blob/5cce4b5f15ffcbfa70079318f284c67de6d8bb6b/app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt#L546.

When the search dialog is open before cliking on the scanner button the url bar is focused and the keyboard is open.
When we click on the scanner button, the keyboard is collapsed 
 - make the permission prompt / permission dialog more visible ?
 - we don't want the keyboard open when the scanner open
 
 If due to the user action we cannot open the scanner it seems logical to me to put back the user in the same initial position with the urlbar selected and keyboard open.
 
 So made some change to the `SearchDialogController` to take a logic specific to the permission dialog to run in all cases.

File: app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt [hunk 2]
```diff
@@ -118,6 +119,14 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
     private var qrButtonAction: Toolbar.Action? = null
     private var voiceSearchButtonAction: Toolbar.Action? = null
 
+    private val cameraPermissionLauncher = registerCameraLauncher(
+        permissionGranted = { qrFeature.get()?.onPermissionsResult(arrayOf(CAMERA), intArrayOf(0)) },
+        permissionDenied = { forever: Boolean ->
+            if( forever ) interactor.onCameraPermissionsNeeded()
+            else resetFocus()
```
Comment (Timshel) [inline app/src/main/java/org/mozilla/fenix/search/SearchDialogFragment.kt:126]: Do not `resetFocus()` when we open the permission dialog to prevent reopening the keyboard while the dialog is open.

**Comment by gabrielluong** [comment]
Hey there, thank you for the contribution! We have migrated the original issue over to Bugzilla https://bugzilla.mozilla.org/show_bug.cgi?id=1815638. When this PR is ready to land, could we please update the commit message and PR title to reference the new Bugzilla issue using the following format: "Bug xxxxxx - \<short description\>". Thanks!

**Comment by csadilek** [comment]
Hey @Timshel, Thank you so much for your contribution. We will be moving this repository to [firefox-android](https://github.com/mozilla-mobile/firefox-android) in the coming days, and won't be able review or land this PR until after  the migration.

Please feel free to reopen a new PR against the new repo once migration is complete. You will have to clone the new repository and reapply your changes. Thank you again, and sorry for the inconvenience.

Since this PR is labelled "Needs UX", it seems it would be best to continue the discussion on the ticket as a next step. /cc @mavduevskiy 

For more information, please see https://github.com/mozilla-mobile/fenix/issues/26855.
